### PR TITLE
Implementation of ESIM model

### PR DESCRIPTION
**Pull request by matt-peters**
A modified version of the ESIM model used in "Deep contextualized word representations" (http://aclweb.org/anthology/N18-1202)

### CODE REVIEW

File: allennlp/models/esim.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+from allennlp.data import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules import FeedForward
+from allennlp.modules.matrix_attention.legacy_matrix_attention import LegacyMatrixAttention
+from allennlp.modules import Seq2SeqEncoder, SimilarityFunction, TextFieldEmbedder
+from allennlp.nn import InitializerApplicator, RegularizerApplicator
+from allennlp.nn.util import get_text_field_mask, last_dim_softmax, weighted_sum, replace_masked_values
+from allennlp.training.metrics import CategoricalAccuracy
+
+class InputVariationalDropout(torch.nn.Dropout):
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/models/esim.py]: done

File: allennlp/models/esim.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    it to every time step.
+    """
+    def forward(self, input_tensor):
+        # pylint: disable=arguments-differ
+        """
+        Apply dropout to input tensor.
+
+        Parameters
+        ----------
+        input_tensor: torch.FloatTensor
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/models/esim.py]: done

File: allennlp/models/esim.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        # The projection layer down to the model dimension.  Dropout is not applied before
+        # projection.
+        projected_enhanced_premise = self._projection_feedforward(premise_enhanced)
+        projected_enhanced_hypothesis = self._projection_feedforward(hypothesis_enhanced)
+
+        # Run the inference layer
+        if self.rnn_input_dropout:
+            projected_enhanced_premise = self.rnn_input_dropout(projected_enhanced_premise)
+            projected_enhanced_hypothesis = self.rnn_input_dropout(projected_enhanced_hypothesis)
+        v_ai = self._inference_encoder(projected_enhanced_premise, premise_mask)
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/models/esim.py]: These names follow the notation in the original paper and are helpful if someone wanted to align the code with equations.

File: allennlp/predictors/esim.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+@Predictor.register('esim')
+class ESIMPredictor(Predictor):
+    """
+    Predictor for the :class:`~allennlp.models.esim.ESIM` model.
+    """
+
+    def predict(self, sentence1: str, sentence2: str) -> JsonDict:
+        """
+        Predicts whether the sentence2 is entailed by the sentence1 text.
```
[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/predictors/esim.py]: Now I remember why I added it -- the decomposable attention predictor uses `premise` and `hypothesis` but the SNLI / MultiNLI data uses `sentence1` and `sentence2`.  This means the existing predictor can't be used with the standard datasets (to e.g. write out predictions to a file for submitting the the leaderboard).  Is there an easy way to over ride the existing tensor names?

[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/predictors/esim.py]: The hacky alternative (to my hacky fix that cuts and pastes the decomposable attention predictor and change the names) is to just modify the incoming json to change the keys, then modify again to back to the original keys.

[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/predictors/esim.py]: @nelson-liu and I were discussing a related issue -- he wanted to use the same model with different dataset readers (which produced differently named fields), we kicked around several bad ideas but never got to any we liked (I don't think)

[the remark this answers is not in the document]
Comment (matt-peters) [reply allennlp/predictors/esim.py]: I'm all for removing the ESIM predictor and will do so now, and open an issue to provide a general solution.

**Pull request merged**